src/cpp/middleware/fast/FastEntities.cpp

```diff
@@ -243,30 +316,32 @@ bool FastDataReader::match_from_xml(const std::string& xml) const
     return rv;
 }
 
-bool FastDataReader::set_on_new_data_cb(OnNewData on_new_data_cb)
-{
-    on_new_data_cb_ = on_new_data_cb;
-    return true;
-}
-
-bool FastDataReader::unset_on_new_data_cb()
-{
-    on_new_data_cb_ = nullptr;
-    return true;
-}
-
-bool FastDataReader::read(std::vector<uint8_t>* data)
+bool FastDataReader::read(std::vector<uint8_t>* data, uint32_t timeout)
 {
+    auto now = std::chrono::steady_clock::now();
     bool rv = false;
+    std::unique_lock<std::mutex> lock(mtx_);
     if (ptr_->getUnreadCount() != 0)
     {
+        lock.unlock();
```

Comment:
If you unlock here, is it means that you can put the `unique_lock` inside of `else` branch and remove this line? (The if condition is not affected by the lock)

Reply:
No, because it is related to the `onNewDataMessage`. Imagine that you call the `getUnreadCount` and it returns 0, then you go through the else but in the meantime, a new message could arrive.